File: packages/node/src/cachelint/hook.ts
/**
 * SDK warn hook: pre-flight cache lint + runtime cache verification.
 *
 * This is the SDK's seat for the ACTUAL rendered outgoing prompt — post
 * routing, post message-assembly — and for the response `usage` fields.
 * The future AST-introspection phase extends this class.
 *
 * Behavior (see docs/sdk/client/cache-lint.md):
 *   - warn-only and fail-open: no method here ever throws — a lint failure
 *     must never break or block the user's API call;
 *   - problems only: warnings fire for BELOW_THRESHOLD,
 *     EFFECTIVELY_UNCACHEABLE, and UNKNOWN_THRESHOLD-with-blockers verdicts;
 *   - once per prompt shape: each unique (model, canonical prompt) pair warns
 *     at most once per client instance (FIFO-capped memory, stale reset);
 *   - verification: from the SECOND call with the same shape onward, a shape
 *     the lint predicted cacheable that reports 0 cached tokens in `usage`
 *     triggers a single VERIFY_MISS warning for that shape.
 *
 * This module is deliberately LIGHT: it has no static value-imports from the
 * engine, so the integrations can import it statically without loading the
 * multi-MB js-tiktoken rank data. The engine loads via a cached dynamic
 * import on the first preflight() only (same discipline as the CLI's
 * `await import('./cachelint/index.js')`).
 *
 * Message formats are byte-identical to the Python SDK's hook (pinned by
 * mirrored unit tests in cachelint-hook.test.ts / test_cachelint_hook.py).
 */

import { createHash } from 'node:crypto';
import { basename } from 'node:path';
import { fileURLToPath } from 'node:url';

type Engine = typeof import('./index.js');
type Introspect = typeof import('./introspect.js');
type CallSite = import('./introspect.js').CallSite;
type OverlaySlot = import('./introspect.js').OverlaySlot;

const PREFIX = '[tryaii cachelint]';
const MAX_SHAPES = 1024; // FIFO cap on remembered prompt shapes
const STALE_AFTER_MS = 3600 * 1000; // idle window after which a shape's counter resets
const MAX_SITES = 512; // FIFO cap on the per-call-site emission/value stores

// The package root for the stack-walk frame filter: '..' from this module is
// src/ under vitest and dist/ when installed — both are exactly what must be
// skipped as SDK-internal.
const PKG_ROOT = fileURLToPath(new URL('..', import.meta.url))
  .replace(/\\/g, '/')
  .toLowerCase();

/** Slot expr for warning lines: whitespace collapsed, capped at 60 chars. */
function displayExpr(expr: string): string {
  const collapsed = expr.replace(/\s+/g, ' ').trim();
  return collapsed.length <= 60 ? collapsed : collapsed.slice(0, 57) + '...';
}

/**
 * Synchronous structured-CallSite capture (spike-verified): V8 zero-cost
 * async CallSites carry the awaiting user frame through internal awaits;
 * node: internal frames (timers, task queues) interleave and are skipped;
 * a first foreign frame inside another node_modules tree is a wrapper
 * library, not user code — dead-end.
 */
function captureCallSite(): CallSite | null {
  const saved = Error.prepareStackTrace;
  const savedLimit = Error.stackTraceLimit;
  let sites: NodeJS.CallSite[];
  try {
    Error.stackTraceLimit = 40;
    Error.prepareStackTrace = (_err, callSites) => callSites;
    sites = new Error().stack as unknown as NodeJS.CallSite[];
  } finally {
    Error.prepareStackTrace = saved;
    Error.stackTraceLimit = savedLimit;
  }
  if (!Array.isArray(sites)) return null;
  for (const site of sites) {
    let file = site.getFileName();
    if (typeof file === 'string' && file.startsWith('file://')) {
      try {
        file = fileURLToPath(file);
      } catch {
        continue;
      }
    }
    if (!file) continue; // anonymous / eval frames
    if (file.startsWith('node:')) continue; // runtime internals interleave
    const norm = file.replace(/\\/g, '/').toLowerCase();
    if (norm.startsWith(PKG_ROOT)) continue; // SDK-internal
    if (norm.includes('/node_modules/')) return null; // wrapper library: dead-end
    const line = site.getLineNumber();
    if (line === null || line === undefined) continue;
    const col = site.getColumnNumber();
    return { file, line, col: col ?? null };
  }
  return null;
}

const NOTE_LINE =
  `${PREFIX} note: reported once per unique prompt shape per client; ` +
  "set cacheLint: 'off' or TRYAII_CACHE_LINT=off to disable";

// Verdicts that warn on preflight. UNKNOWN_THRESHOLD additionally requires a
// blocking finding: bare UNKNOWN is a knowledge-base gap (e.g. a deepseek
// slug), not a problem in the user's prompt.
const WARN_VERDICTS = new Set(['BELOW_THRESHOLD', 'EFFECTIVELY_UNCACHEABLE', 'UNKNOWN_THRESHOLD']);

// Verdicts under which a cache read is genuinely expected on repeat calls.
// CACHEABLE_WITH_ACTION is excluded: this SDK never sends cache_control, so
// zero cached tokens is the CORRECT outcome there, not a mismatch.
const EXPECT_CACHE_VERDICTS = new Set(['CACHEABLE', 'CACHEABLE_PREFIX']);

interface Prediction {
  model: string;
  verdict: string;
  stableTokens: number;
  minTokens: number | null;
  expectCache: boolean;
  calls: number;
  verifyWarned: boolean;
  lastSeen: number;
  templateSlots: OverlaySlot[] | null;
}

export interface CacheLintHookOptions {
  /** Warning line sink; defaults to process.stderr. */
  sink?: (line: string) => void;
  /** Monotonic-ish clock in milliseconds; defaults to Date.now. */
  now?: () => number;
  /** Engine loader test seam; defaults to importing the cachelint module. */
  loadEngine?: () => Promise<Engine>;
  /** Introspection loader test seam; defaults to importing the AST module. */
  loadIntrospect?: () => Promise<Introspect>;
}

/**
 * Resolve the cacheLint mode and build the hook (or null when off).
 * Precedence: explicit option value > TRYAII_CACHE_LINT env var > off.
 * Ambient/env garbage is treated as off — it must never throw.
 */
export function buildCacheLintHook(value?: string): CacheLintHook | null {
  const mode = value ?? process.env.TRYAII_CACHE_LINT ?? '';
  return mode.trim().toLowerCase() === 'warn' ? new CacheLintHook() : null;
}

/** Per-client-instance lint state. Public methods never throw. */
export class CacheLintHook {
  private readonly _sink: (line: string) => void;
  private readonly _now: () => number;
  private readonly _loadEngine: () => Promise<Engine>;
  private _engine: Promise<Engine> | null = null;
  private _engineBroken = false;
  private readonly _seen = new Map<string, Prediction>();
  private _noteShown = false;
  private readonly _loadIntrospect: () => Promise<Introspect>;
  private _introspect: Promise<Introspect> | null = null;
  private _introspectBroken = false;
  // AST-introspection emission state (once per call-site+slot, FIFO-capped).
  private readonly _siteWarned = new Map<string, null>();
  private readonly _slotValues = new Map<string, string>();

  constructor(options?: CacheLintHookOptions) {
    this._sink = options?.sink ?? ((line: string) => process.stderr.write(line + '\n'));
    this._now = options?.now ?? (() => Date.now());
    this._loadEngine = options?.loadEngine ?? (() => import('./index.js'));
    this._loadIntrospect = options?.loadIntrospect ?? (() => import('./introspect.js'));
  }

  private _emit(line: string): void {
    try {
      this._sink(line);
    } catch {
      // a broken sink must not break the call
    }
  }

  private _key(openrouterModel: string, canonical: string): string {
    return createHash('sha256')
      .update(openrouterModel + '\x00' + canonical, 'utf-8')
      .digest('hex')
      .slice(0, 16);
  }

  /**
   * Lint the outgoing request; resolves to the shape key (or null on failure).
   * `callSite` is an injected test seam; when absent the hook captures the
   * stack SYNCHRONOUSLY at entry (before any await) for AST introspection.
   */
  async preflight(
    openrouterModel: string,
    messages: Array<{ role: string; content: string }>,
    callSite?: CallSite | null,
  ): Promise<string | null> {
    let site: CallSite | null = callSite ?? null;
    if (site === null && callSite === undefined) {
      try {
        site = captureCallSite();
      } catch {
        site = null;
      }
    }
    try {
      return await this._preflight(openrouterModel, messages, site);
    } catch {
      return null; // fail-open: never break the user's call
    }
  }

  private async _preflight(
    openrouterModel: string,
    messages: Array<{ role: string; content: string }>,
    site: CallSite | null,
  ): Promise<string | null> {
    if (this._engineBroken) return null;
    this._engine ??= this._loadEngine();
    let engine: Engine;
    try {
      engine = await this._engine;
    } catch {
      this._engineBroken = true; // permanent no-op; never retry a broken load
      return null;
    }

    const [canonical, sections] = engine.buildCanonical({ messages });
    const key = this._key(openrouterModel, canonical);
    const now = this._now();

    const seen = this._seen.get(key);
    if (seen !== undefined) {
      if (now - seen.lastSeen > STALE_AFTER_MS) seen.calls = 0;
      seen.calls++;
      seen.lastSeen = now;
      return key;
    }

    if (this._seen.size >= MAX_SHAPES) {
      const oldest = this._seen.keys().next().value as string | undefined;
      if (oldest !== undefined) this._seen.delete(oldest);
    }

    const result = engine.analyze({
      prompt: { messages },
      llm: { provider: 'openrouter', name: openrouterModel },
    }) as { items: Array<Record<string, any>> };
    const item = result.items[0];
    const verdict = item.verdict as { code: string; summary: string };
    const limitedBy = item.stable_prefix.limited_by as Record<string, any> | null;

    // Concurrent first calls with the same shape can interleave across the
    // awaits above — keep the earlier record if one landed meanwhile.
    const raced = this._seen.get(key);
    if (raced !== undefined) {
      raced.calls++;
      raced.lastSeen = now;
      return key;
    }

    const prediction: Prediction = {
      model: openrouterModel,
      verdict: verdict.code,
      stableTokens: item.stable_prefix.tokens as number,
      minTokens: item.threshold.min_tokens as number | null,
      expectCache: EXPECT_CACHE_VERDICTS.has(verdict.code),
      calls: 1,
      verifyWarned: false,
      lastSeen: now,
      templateSlots: null,
    };
    this._seen.set(key, prediction);

    // AST template introspection (fail-open; new shapes only by construction).
    const insights = await this._templateStep(messages, canonical, sections, item, site);
    if (insights) {
      prediction.templateSlots = insights.flatMap(([, classified]) =>
        classified.map(([slot]) => slot));
    }

    const triggered =
      WARN_VERDICTS.has(verdict.code) &&
      (verdict.code !== 'UNKNOWN_THRESHOLD' || limitedBy !== null);
    let emitted = false;
    if (triggered) {
      emitted = true;
      this._emit(`${PREFIX} ${verdict.code} for ${openrouterModel}: ${verdict.summary}`);
      if (limitedBy !== null) {
        this._emit(
          `${PREFIX}   first blocker: ${limitedBy.kind} at ` +
            `${limitedBy.section}+${limitedBy.section_offset} ` +
            `(${Math.floor(limitedBy.pct_into_prompt)}% in)`,
        );
        if (insights) this._emitAttribution(insights, limitedBy);
      }
    } else if (insights) {
      // Only today-silent verdicts reach here (BELOW_THRESHOLD always
      // triggers above) — the standalone slot lines fill the blind spot.
      emitted = this._emitSlotWarnings(insights);
    }
    if (emitted && !this._noteShown) {
      this._noteShown = true;
      this._emit(NOTE_LINE);
    }
    return key;
  }

  // -- AST template introspection --------------------------------------

  private async _templateStep(
    messages: Array<{ role: string; content: string }>,
    canonical: string,
    sections: Array<{ name: string; start: number; end: number }>,
    item: Record<string, any>,
    site: CallSite | null,
  ): Promise<Array<[CallSite, Array<[OverlaySlot, boolean]>]> | null> {
    try {
      if (site === null || this._introspectBroken) return null;
      this._introspect ??= this._loadIntrospect();
      let introspect: Introspect;
      try {
        introspect = await this._introspect;
      } catch {
        this._introspectBroken = true;
        return null;
      }

      const maps = await introspect.analyzeCallSite(site, messages);
      if (!maps) return null;
      const limitedBy = item.stable_prefix.limited_by as Record<string, any> | null;
      const boundary = limitedBy !== null ? (limitedBy.start as number) : canonical.length;

      const out: Array<[CallSite, Array<[OverlaySlot, boolean]>]> = [];
      for (const tmap of maps) {
        const idx = this._targetMessageIndex(tmap.target, messages);
        if (idx === null || idx >= sections.length) continue;
        const role = messages[idx].role ?? 'user';
        const rendered = messages[idx].content;
        if (typeof rendered !== 'string') continue;
        const spans = introspect.align(tmap.segments, rendered);
        if (spans === null) continue; // validation guard: discard the analysis
        const contentStart = sections[idx].start + role.length + 2;
        const slots = introspect.overlay(tmap.segments, spans, contentStart, boundary);
        const classified: Array<[OverlaySlot, boolean]> = [];
        for (const slot of slots) {
          const valueKey = `${site.file}:${site.line}:${slot.expr}`;
          const value = canonical.slice(slot.start, slot.end);
          const prev = this._slotValues.get(valueKey);
          if (this._slotValues.size >= MAX_SITES && !this._slotValues.has(valueKey)) {
            const oldest = this._slotValues.keys().next().value as string | undefined;
            if (oldest !== undefined) this._slotValues.delete(oldest);
          }
          this._slotValues.set(valueKey, value);
          const dynamic = slot.hasCall || (prev !== undefined && prev !== value);
          classified.push([slot, dynamic]);
        }
        if (classified.length) out.push([site, classified]);
      }
      return out.length ? out : null;
    } catch {
      return null; // fail-open: introspection is a bonus
    }
  }

  private _targetMessageIndex(
    target: string,
    messages: Array<{ role: string; content: string }>,
  ): number | null {
    const role = target === 'system_message' ? 'system' : 'user';
    for (let i = messages.length - 1; i >= 0; i--) {
      if (messages[i] && messages[i].role === role) return i;
    }
    return null;
  }

  private _emitSlotWarnings(
    insights: Array<[CallSite, Array<[OverlaySlot, boolean]>]>,
  ): boolean {
    let shown = 0;
    let extra = 0;
    for (const [site, classified] of insights) {
      const base = basename(site.file);
      for (const [slot, dynamic] of classified) {
        if (!slot.inPrefix || !dynamic) continue;
        const warnKey = `${site.file}:${site.line}:${slot.expr}`;
        if (this._siteWarned.has(warnKey)) continue;
        if (this._siteWarned.size >= MAX_SITES) {
          const oldest = this._siteWarned.keys().next().value as string | undefined;
          if (oldest !== undefined) this._siteWarned.delete(oldest);
        }
        this._siteWarned.set(warnKey, null);
        if (shown >= 3) {
          extra++;
          continue;
        }
        shown++;
        const expr = displayExpr(slot.expr);
        this._emit(
          `${PREFIX} template slot {${expr}} at ${base}:${site.line} ` +
            'renders inside your cacheable prefix — its value changes ' +
            'between calls and breaks the cache there',
        );
        if (slot.resolvedFrom !== null) {
          const fromExpr = displayExpr(slot.resolvedFrom[0]);
          this._emit(`${PREFIX}   {${expr}} = ${fromExpr} at ${base}:${slot.resolvedFrom[1]}`);
        }
      }
    }
    if (extra) {
      this._emit(`${PREFIX}   ...and ${extra} more template slot(s) in the cacheable prefix`);
    }
    return shown > 0 || extra > 0;
  }

  private _emitAttribution(
    insights: Array<[CallSite, Array<[OverlaySlot, boolean]>]>,
    limitedBy: Record<string, any>,
  ): void {
    const bStart = limitedBy.start as number;
    const bEnd = limitedBy.end as number;
    for (const [site, classified] of insights) {
      for (const [slot] of classified) {
        if (slot.start < bEnd && bStart < slot.end) {
          const warnKey = `attr:${site.file}:${site.line}:${slot.expr}`;
          if (this._siteWarned.has(warnKey)) return;
          if (this._siteWarned.size >= MAX_SITES) {
            const oldest = this._siteWarned.keys().next().value as string | undefined;
            if (oldest !== undefined) this._siteWarned.delete(oldest);
          }
          this._siteWarned.set(warnKey, null);
          const base = basename(site.file);
          const expr = displayExpr(slot.expr);
          this._emit(`${PREFIX}   rendered by template slot {${expr}} at ${base}:${site.line}`);
          return; // at most one attribution line per warn block
        }
      }
    }
  }

  /** Compare the preflight prediction against the response usage fields. */
  verify(key: string | null, usage: unknown): void {
    try {
      this._verify(key, usage);
    } catch {
      // fail-open: never break the user's call
    }
  }

  private _verify(key: string | null, usage: unknown): void {
    if (key === null || usage === null || usage === undefined) return;
    if (typeof usage !== 'object' || Array.isArray(usage)) return;
    const u = usage as Record<string, unknown>;
    const rec = this._seen.get(key);
    if (rec === undefined || rec.calls < 2 || !rec.expectCache || rec.verifyWarned) return;

    let cached: number | null = null;
    const details = u.prompt_tokens_details;
    if (
      details !== null &&
      typeof details === 'object' &&
      typeof (details as Record<string, unknown>).cached_tokens === 'number'
    ) {
      cached = (details as Record<string, unknown>).cached_tokens as number;
    } else if (typeof u.cached_tokens === 'number') {
      cached = u.cached_tokens;
    }

    const discount = typeof u.cache_discount === 'number' ? u.cache_discount : null;

    if (cached === null && discount === null) return; // unverifiable — stay silent
    if ((cached ?? 0) === 0 && (discount ?? 0) === 0) {
      rec.verifyWarned = true;
      this._emit(
        `${PREFIX} VERIFY_MISS for ${rec.model}: predicted ~${rec.stableTokens} tok ` +
          `cacheable prefix, but usage reports 0 cached tokens on repeat call ` +
          `#${rec.calls} of this prompt shape — a silent invalidator or missing ` +
          `provider support may be the cause`,
      );
    }
  }
}

File: packages/python/tryaii/cachelint/hook.py
"""SDK warn hook: pre-flight cache lint + runtime cache verification.

This is the SDK's seat for the ACTUAL rendered outgoing prompt — post
routing, post message-assembly, post truncation — and for the response
`usage` fields. The future AST-introspection phase extends this class.

Behavior (see docs/sdk/client/cache-lint.md):
  * warn-only and fail-open: no method here ever raises — a lint failure
    must never break or block the user's API call;
  * problems only: warnings fire for BELOW_THRESHOLD, EFFECTIVELY_UNCACHEABLE,
    and UNKNOWN_THRESHOLD-with-blockers verdicts — never for healthy prompts;
  * once per prompt shape: each unique (model, canonical prompt) pair warns at
    most once per client instance (FIFO-capped memory, stale entries reset);
  * verification: from the SECOND call with the same shape onward, a prompt
    the lint predicted cacheable that reports 0 cached tokens in `usage`
    triggers a single VERIFY_MISS warning for that shape.

Output goes to stderr by default: the package root installs a NullHandler,
so `logger.warning` would be invisible — and a user who explicitly opted in
with cache_lint="warn" has asked for visible output. This is deliberately
the only library-level stderr writer outside the CLI, and it only ever runs
under that explicit opt-in. Each warning is mirrored to logger.debug for
apps with logging configured.

Message formats are byte-identical to the Node SDK's hook (pinned by
mirrored unit tests in test_cachelint_hook.py / cachelint-hook.test.ts).
"""

from __future__ import annotations

import hashlib
import logging
import os
import re
import sys
import time
from dataclasses import dataclass
from typing import Callable, Optional

logger = logging.getLogger("tryaii.cachelint.hook")

_PREFIX = "[tryaii cachelint]"
_MAX_SHAPES = 1024          # FIFO cap on remembered prompt shapes
_STALE_AFTER_S = 3600.0     # idle window after which a shape's call counter resets

_NOTE_LINE = (
    f"{_PREFIX} note: reported once per unique prompt shape per client; "
    'set cache_lint="off" or TRYAII_CACHE_LINT=off to disable'
)
_TIKTOKEN_HINT = (
    f"{_PREFIX} note: cache lint skipped for an OpenAI/xAI-routed prompt — "
    "install tryaii[cachelint] to enable exact token analysis"
)

# Verdicts that warn on preflight. UNKNOWN_THRESHOLD additionally requires a
# blocking finding: bare UNKNOWN is a knowledge-base gap (e.g. a deepseek
# slug), not a problem in the user's prompt.
_WARN_VERDICTS = ("BELOW_THRESHOLD", "EFFECTIVELY_UNCACHEABLE", "UNKNOWN_THRESHOLD")

# Verdicts under which a cache read is genuinely expected on repeat calls.
# CACHEABLE_WITH_ACTION is excluded: this SDK never sends cache_control, so
# zero cached tokens is the CORRECT outcome there, not a mismatch.
_EXPECT_CACHE_VERDICTS = ("CACHEABLE", "CACHEABLE_PREFIX")


_MAX_SITES = 512   # FIFO cap on the per-call-site emission/value stores


def _stderr_sink(line: str) -> None:
    # Resolved at call time so monkeypatched/captured stderr is honored.
    sys.stderr.write(line + "\n")


def _display_expr(expr: str) -> str:
    """Slot expr for warning lines: whitespace collapsed, capped at 60 chars."""
    collapsed = re.sub(r"\s+", " ", expr).strip()
    return collapsed if len(collapsed) <= 60 else collapsed[:57] + "..."


@dataclass
class _Prediction:
    model: str
    verdict: str
    stable_tokens: int
    min_tokens: Optional[int]
    expect_cache: bool
    calls: int
    verify_warned: bool
    last_seen: float
    template_slots: Optional[list] = None   # OverlaySlots from AST introspection


class CacheLintHook:
    """Per-client-instance lint state. Public methods never raise."""

    def __init__(
        self,
        sink: Optional[Callable[[str], None]] = None,
        clock: Callable[[], float] = time.monotonic,
    ):
        self._sink = sink or _stderr_sink
        self._clock = clock
        self._seen: dict[str, _Prediction] = {}
        self._note_shown = False
        self._tiktoken_hint_shown = False
        # AST-introspection emission state (once per call-site+slot, FIFO-capped).
        self._site_warned: dict[str, None] = {}
        self._slot_values: dict[str, str] = {}

    # -- internals ------------------------------------------------------

    def _emit(self, line: str) -> None:
        try:
            self._sink(line)
        except Exception:  # noqa: BLE001 -- a broken sink must not break the call
            pass
        logger.debug("%s", line)

    @staticmethod
    def _key(openrouter_model: str, canonical: str) -> str:
        raw = (openrouter_model + "\x00" + canonical).encode("utf-8")
        return hashlib.sha256(raw).hexdigest()[:16]

    # -- public API -----------------------------------------------------

    def preflight(self, openrouter_model: str, messages: list,
                  call_site=None) -> Optional[str]:
        """Lint the outgoing request; returns the shape key (or None on failure).

        `call_site` (an _introspect.CallSite) is an injected test seam; when
        None the hook walks the stack itself for AST template introspection.
        """
        try:
            return self._preflight(openrouter_model, messages, call_site)
        except ImportError:
            if not self._tiktoken_hint_shown:
                self._tiktoken_hint_shown = True
                self._emit(_TIKTOKEN_HINT)
            return None
        except Exception:  # noqa: BLE001 -- fail-open: never break the user's call
            logger.debug("cachelint preflight error", exc_info=True)
            return None

    def _preflight(self, openrouter_model: str, messages: list,
                   call_site=None) -> Optional[str]:
        from tryaii.cachelint import analyze, build_canonical

        canonical, sections, _ = build_canonical({"messages": messages})
        key = self._key(openrouter_model, canonical)
        now = self._clock()

        rec = self._seen.get(key)
        if rec is not None:
            if now - rec.last_seen > _STALE_AFTER_S:
                rec.calls = 0
            rec.calls += 1
            rec.last_seen = now
            return key

        if len(self._seen) >= _MAX_SHAPES:
            self._seen.pop(next(iter(self._seen)))

        result = analyze({
            "prompt": {"messages": messages},
            "llm": {"provider": "openrouter", "name": openrouter_model},
        })
        item = result["items"][0]
        verdict = item["verdict"]
        limited_by = item["stable_prefix"]["limited_by"]

        prediction = _Prediction(
            model=openrouter_model,
            verdict=verdict["code"],
            stable_tokens=item["stable_prefix"]["tokens"],
            min_tokens=item["threshold"]["min_tokens"],
            expect_cache=verdict["code"] in _EXPECT_CACHE_VERDICTS,
            calls=1,
            verify_warned=False,
            last_seen=now,
        )
        self._seen[key] = prediction

        # AST template introspection (fail-open; new shapes only by construction).
        insights = self._template_step(messages, canonical, sections, item, call_site)
        if insights:
            prediction.template_slots = [slot for _, slots in insights for slot in slots]

        triggered = verdict["code"] in _WARN_VERDICTS and (
            verdict["code"] != "UNKNOWN_THRESHOLD" or limited_by is not None
        )
        emitted = False
        if triggered:
            emitted = True
            self._emit(f"{_PREFIX} {verdict['code']} for {openrouter_model}: "
                       f"{verdict['summary']}")
            if limited_by is not None:
                self._emit(f"{_PREFIX}   first blocker: {limited_by['kind']} at "
                           f"{limited_by['section']}+{limited_by['section_offset']} "
                           f"({int(limited_by['pct_into_prompt'])}% in)")
                if insights:
                    self._emit_attribution(insights, limited_by)
        elif insights:
            # Only today-silent verdicts reach here (BELOW_THRESHOLD always
            # triggers above) — the standalone slot lines fill the blind spot.
            emitted = self._emit_slot_warnings(insights)
        if emitted and not self._note_shown:
            self._note_shown = True
            self._emit(_NOTE_LINE)
        return key

    # -- AST template introspection --------------------------------------

    def _template_step(self, messages: list, canonical: str, sections: list,
                       item: dict, call_site) -> Optional[list]:
        """Trace, align, overlay, and classify template slots. Fail-open.

        Returns [(site, [(slot, dynamic_bool), ...])] or None. Also arms the
        tier-1 value store for every slot (a bare-name slot warns on the call
        that PROVES it dynamic — i.e. when its rendered value changes).
        """
        try:
            from tryaii.cachelint import _introspect

            site = call_site if call_site is not None else _introspect.capture_call_site()
            if site is None:
                return None
            maps = _introspect.analyze_call_site(site, messages)
            if not maps:
                return None
            limited_by = item["stable_prefix"]["limited_by"]
            boundary = limited_by["start"] if limited_by is not None else len(canonical)

            out = []
            for tmap in maps:
                idx = self._target_message_index(tmap.target, messages)
                if idx is None or idx >= len(sections):
                    continue
                role = messages[idx].get("role", "user")
                rendered = messages[idx].get("content", "")
                if not isinstance(rendered, str):
                    continue
                spans = _introspect.align(tmap.segments, rendered)
                if spans is None:
                    continue          # validation guard: discard the analysis
                content_start = sections[idx].start + len(role) + 2
                slots = _introspect.overlay(tmap.segments, spans, content_start, boundary)
                classified = []
                for slot in slots:
                    value_key = f"{site.file}:{site.line}:{slot.expr}"
                    value = canonical[slot.start:slot.end]
                    prev = self._slot_values.get(value_key)
                    if len(self._slot_values) >= _MAX_SITES and value_key not in self._slot_values:
                        self._slot_values.pop(next(iter(self._slot_values)))
                    self._slot_values[value_key] = value
                    dynamic = slot.has_call or (prev is not None and prev != value)
                    classified.append((slot, dynamic))
                if classified:
                    out.append((site, classified))
            return out or None
        except Exception:  # noqa: BLE001 -- fail-open: introspection is a bonus
            logger.debug("cachelint introspection error", exc_info=True)
            return None

    @staticmethod
    def _target_message_index(target: str, messages: list) -> Optional[int]:
        role = "system" if target == "system_message" else "user"
        for i in range(len(messages) - 1, -1, -1):
            if isinstance(messages[i], dict) and messages[i].get("role") == role:
                return i
        return None

    def _emit_slot_warnings(self, insights: list) -> bool:
        shown = 0
        extra = 0
        for site, classified in insights:
            base = os.path.basename(site.file)
            for slot, dynamic in classified:
                if not slot.in_prefix or not dynamic:
                    continue
                warn_key = f"{site.file}:{site.line}:{slot.expr}"
                if warn_key in self._site_warned:
                    continue
                if len(self._site_warned) >= _MAX_SITES:
                    self._site_warned.pop(next(iter(self._site_warned)))
                self._site_warned[warn_key] = None
                if shown >= 3:
                    extra += 1
                    continue
                shown += 1
                expr = _display_expr(slot.expr)
                self._emit(f"{_PREFIX} template slot {{{expr}}} at {base}:{site.line} "
                           "renders inside your cacheable prefix — its value changes "
                           "between calls and breaks the cache there")
                if slot.resolved_from is not None:
                    from_expr = _display_expr(slot.resolved_from[0])
                    self._emit(f"{_PREFIX}   {{{expr}}} = {from_expr} at "
                               f"{base}:{slot.resolved_from[1]}")
        if extra:
            self._emit(f"{_PREFIX}   ...and {extra} more template slot(s) in the "
                       "cacheable prefix")
        return shown > 0 or extra > 0

    def _emit_attribution(self, insights: list, limited_by: dict) -> None:
        b_start, b_end = limited_by["start"], limited_by["end"]
        for site, classified in insights:
            for slot, _dynamic in classified:
                if slot.start < b_end and b_start < slot.end:   # spans intersect
                    warn_key = f"attr:{site.file}:{site.line}:{slot.expr}"
                    if warn_key in self._site_warned:
                        return
                    if len(self._site_warned) >= _MAX_SITES:
                        self._site_warned.pop(next(iter(self._site_warned)))
                    self._site_warned[warn_key] = None
                    base = os.path.basename(site.file)
                    expr = _display_expr(slot.expr)
                    self._emit(f"{_PREFIX}   rendered by template slot {{{expr}}} "
                               f"at {base}:{site.line}")
                    return   # at most one attribution line per warn block

    def verify(self, key: Optional[str], usage) -> None:
        """Compare the preflight prediction against the response usage fields."""
        try:
            self._verify(key, usage)
        except Exception:  # noqa: BLE001 -- fail-open: never break the user's call
            logger.debug("cachelint verify error", exc_info=True)

    def _verify(self, key: Optional[str], usage) -> None:
        if key is None or not isinstance(usage, dict):
            return
        rec = self._seen.get(key)
        if rec is None or rec.calls < 2 or not rec.expect_cache or rec.verify_warned:
            return

        cached = None
        details = usage.get("prompt_tokens_details")
        if isinstance(details, dict) and isinstance(details.get("cached_tokens"), (int, float)) \
                and not isinstance(details.get("cached_tokens"), bool):
            cached = details["cached_tokens"]
        elif isinstance(usage.get("cached_tokens"), (int, float)) \
                and not isinstance(usage.get("cached_tokens"), bool):
            cached = usage["cached_tokens"]

        discount = usage.get("cache_discount")
        if not isinstance(discount, (int, float)) or isinstance(discount, bool):
            discount = None

        if cached is None and discount is None:
            return  # unverifiable — stay silent
        if (cached or 0) == 0 and (discount or 0) == 0:
            rec.verify_warned = True
            self._emit(
                f"{_PREFIX} VERIFY_MISS for {rec.model}: predicted ~{rec.stable_tokens} tok "
                f"cacheable prefix, but usage reports 0 cached tokens on repeat call "
                f"#{rec.calls} of this prompt shape — a silent invalidator or missing "
                f"provider support may be the cause"
            )
